storage/jellyfish_merkle/src/restore/mod.rs

```diff
@@ -0,0 +1,426 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module implements the functionality to restore a `JellyfishMerkleTree` from small chunks
+//! of accounts.
+
+#[cfg(test)]
+mod restore_test;
+
+use crate::{
+    nibble::NibblePath,
+    node_type::{Child, Children, InternalNode, LeafNode, Node, NodeKey},
+    NodeBatch, TreeReader, TreeWriter, ROOT_NIBBLE_HEIGHT,
+};
+use crypto::{hash::CryptoHash, HashValue};
+use failure::prelude::*;
+use types::{account_state_blob::AccountStateBlob, transaction::Version};
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+enum ChildInfo {
+    /// This child is an internal node. The hash of the internal node is stored here if it is
+    /// known, otherwise it is `None`. In the process of restoring a tree, we will only know the
+    /// hash of an internal node after we see all the keys that share the same prefix.
+    Internal { hash: Option<HashValue> },
+
+    /// This child is a leaf node.
+    Leaf { node: LeafNode },
+}
+
+#[derive(Clone, Debug)]
+struct InternalInfo {
+    /// The node key of this internal node.
+    node_key: NodeKey,
+
+    /// The existing children. Every time a child appears, the corresponding position will be set
+    /// to `Some`.
+    children: [Option<ChildInfo>; 16],
```

Comment:
I kinda think it's more intuitive to use a `HashMap` as a jellyfish internal node does.
Is this for performance consideration? (what's the difference between this and the jellyfish internal node case?)

Reply:
I actually don't like the HashMap in jellyfish internal node, especially due to the way it serializes.  I was thinking of changing or getting rid of it.